pkg/location/reference.go

```diff
@@ -0,0 +1,13 @@
+package location
+
+import "fmt"
+
+type Reference interface {
+	fmt.Stringer
+	Type() string
+	Validate() error
```

Comment:
Is Validate needed given that the underlying type can ensure that invalid instance cannot exist?

Or are there cases where otherwise valid reference moves through invalid state(s)?

Reply:
Well, I suppose a short answer would be that `Validate()` was added to be called from places in kpt commands that call validate on arguments...

Beyond that though - the types like `location.Dir{}` being simple struct with public fields for convenience does mean the calling code can initialize or modify the values as needed.

So in the end that would be the case where invalid an ref can be created pretty easily, and calling Validate at the point of consumption would be the guarantee.